### PR TITLE
Major overhaul of API, implemented Async compatible repeat/until_* nodes

### PR DESCRIPTION
**Pull request by jbcpollak**

### CODE REVIEW

File: common/decorator/until_failure_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	go func() {
+		tree.EventLoop(evt)
+		wg.Done()
+	}()
+
+	util.PrintTreeInColor(tree.Root)
+
+	d := time.Duration(100) * time.Millisecond
+
+	for loop := true; loop; {
```
Comment (nivanov-ati) [inline common/decorator/until_failure_test.go:67]: I think it's more canonical to do an infinite while loop here, like:

for {
....
}

and just do "break" on line 74

File: common/decorator/until_success_test.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	go func() {
+		tree.EventLoop(evt)
+		wg.Done()
+	}()
+
+	util.PrintTreeInColor(tree.Root)
+
+	d := time.Duration(100) * time.Millisecond
+
+	for loop := true; loop; {
```
Comment (nivanov-ati) [inline common/decorator/until_success_test.go:67]: same comment

**Review by nivanov-ati — CHANGES_REQUESTED** [review]

File: common/decorator/repeat_until.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	return &repeatUntil[Blackboard]{
+		Decorator: base,
+		until:     params.Until,
+	}
+}
+
+type repeatUntil[Blackboard any] struct {
+	core.Decorator[Blackboard, RepeatUntilParams]
+
+	until UntilCondition
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/decorator/repeat_until.go:30]: good point

File: common/decorator/until_failure_test.go (hunk 1 above)
Comment (jbcpollak) [reply common/decorator/until_failure_test.go:67]: I started with that - the break breaks out of the select, not the for, AFAIK this is the only way to do this. 

@@fastcat is there a better way?

Comment (nivanov-ati) [reply common/decorator/until_failure_test.go:67]: According to the tutorial, go doesnt need "break" in switch case statements, but I'm not sure if using one anyways would make it break out of `select` or `for`. It sounds like you tested it and it gets consumed by the select, so feel free to keep it as is

File: common/action/counter.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+type CounterParams struct {
+	core.BaseParams
+
+	Limit     uint
+	CountChan chan uint
+}
+
+// Counter increments a counter on the blackboard until it reaches a certain value.
+func Counter[Blackboard any](params CounterParams) core.Node[Blackboard] {
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/action/counter.go:18]: probably not, but its a good example node and I figured hiding in a test-only directory would be more trouble than its worth.

File: common/decorator/until_failure_test.go (hunk 1 above)
Comment (fastcat) [reply common/decorator/until_failure_test.go:67]: Use a label, I'll make a suggestion (requires a separate thread)

**Review by nivanov-ati — APPROVED** [review]

File: common/decorator/until_failure_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		case c := <-countChan:
+			log.Info().Msgf("got count %v", c)
+		case c := <-sigChan:
+			log.Info().Msgf("loop is finished %v", c)
+
+			loop = false
+		case <-time.After(d):
+			t.Errorf("Timeout after delaying %v", d)
+		}
+	}
```
Comment (fastcat) [inline common/decorator/until_failure_test.go:78]:
```suggestion
	LOOP: for {
		select {
		case c := <-countChan:
			log.Info().Msgf("got count %v", c)
		case c := <-sigChan:
			log.Info().Msgf("loop is finished %v", c)
			break LOOP
		case <-time.After(d):
			t.Errorf("Timeout after delaying %v", d)
		}
	}
```
(gofmt will move the label a bit, I wrote it this way to make the diff easier to view on GH)

**Pull request merged**